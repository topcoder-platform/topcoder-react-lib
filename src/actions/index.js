import authActions from './auth';
import smpActions from './smp';
import statsActions from './stats';
import termsActions from './terms';
import directActions from './direct';
import groupActions from './groups';
import errorActions from './errors';
import challengeActions from './challenge';
import profileActions from './profile';
import memberActions from './members';
import memberTaskActions from './member-tasks';
import reviewOpportunityActions from './reviewOpportunity';
import lookupActions from './lookup';

export const actions = {
  auth: authActions.auth,
  smp: smpActions.smp,
  challenge: challengeActions.challenge,
  terms: termsActions.terms,
  stats: statsActions.stats,
  direct: directActions.direct,
  groups: groupActions.groups,
  errors: errorActions.errors,
  profile: profileActions.profile,
  members: memberActions.members,
  memberTasks: memberTaskActions.memberTasks,
  reviewOpportunity: reviewOpportunityActions.reviewOpportunity,
  lookup: lookupActions.lookup,
};

export default undefined;
