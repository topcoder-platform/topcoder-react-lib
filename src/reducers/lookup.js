/**
 * @module "reducers.lookup"
 * @desc Reducer for {@link module:actions.lookup} actions.
 *
 * State segment managed by this reducer has the following structure:
 * @param {Array} skillTags='' skill tags.
 */
import _ from 'lodash';
import { handleActions } from 'redux-actions';
import logger from '../utils/logger';
import actions from '../actions/lookup';

/**
 * Handles LOOKUP/GET_SKILL_TAGS_DONE action.
 * @param {Object} state
 * @param {Object} action Payload will be JSON from api call
 * @return {Object} New state
 */
function onGetSkillTagsDone(state, { payload, error }) {
  if (error) {
    logger.error('Failed to get skill tags', payload);
    return { ...state, loadingSkillTagsError: true };
  }

  return ({
    ...state,
    loadingSkillTagsError: false,
    skillTags: payload,
  });
}

/**
 * Creates a new Lookup reducer with the specified initial state.
 * @param {Object} initialState Optional. Initial state.
 * @return {Function} Lookup reducer.
 */
function create(initialState = {}) {
  const a = actions.lookup;
  return handleActions({
    [a.getSkillTagsInit]: state => state,
    [a.getSkillTagsDone]: onGetSkillTagsDone,
  }, _.defaults(initialState, {
    skillTags: [],
  }));
}

/**
 * Factory which creates a new reducer.
 * @return {Promise}
 * @resolves {Function(state, action): state} New reducer.
 */
export function factory() {
  return Promise.resolve(create());
}

/**
 * @static
 * @member default
 * @desc Reducer with default initial state.
 */
export default create();
