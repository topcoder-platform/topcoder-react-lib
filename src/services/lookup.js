/**
 * @module "services.lookup"
 * @desc  This module provides a service to get lookup data from Topcoder
 * via API V3.
 */
import qs from 'qs';
import { getApiResponsePayload } from '../utils/tc';
import { getApi } from './api';

class LookupService {
  /**
   * @param {String} tokenV3 Optional. Auth token for Topcoder API v3.
   */
  constructor(tokenV3) {
    this.private = {
      api: getApi('V3', tokenV3),
      apiV5: getApi('V5', tokenV3),
      tokenV3,
    };
  }

  /**
   * Gets tags.
   * @param {Object} params Parameters
   * @return {Promise} Resolves to the tags.
   */
  async getTags(params) {
    const res = await this.private.api.get(`/tags/?${qs.stringify(params)}`);
    return getApiResponsePayload(res);
  }

  /**
   * Gets countries.
   * @param {Object} params Parameters
   * @return {Promise} Resolves to the countries.
   */
  async getCountries() {
    const res = await this.private.api.get('/members/lookup/countries');
    return getApiResponsePayload(res);
  }

  /**
   * Gets all countries.
   * @return {Promise} Resolves to the countries.
   */
  async getAllCountries() {
    const res = await this.private.apiV5.get('/lookups/countries');
    const jsonResult = await res.json();
    return jsonResult;
  }
}

let lastInstance = null;
/**
 * Returns a new or existing lookup service.
 * @param {String} tokenV3 Optional. Auth token for Topcoder API v3.
 * @return {LookupService} Lookup service object
 */
export function getService(tokenV3) {
  if (!lastInstance || tokenV3 !== lastInstance.private.tokenV3) {
    lastInstance = new LookupService(tokenV3);
  }
  return lastInstance;
}

/* Using default export would be confusing in this case. */
export default undefined;
