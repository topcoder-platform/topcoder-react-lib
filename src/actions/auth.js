/**
 * @module "actions.auth"
 * @desc Actions related to Topcoder authentication system.
 */

import { createActions } from 'redux-actions';
import { decodeToken } from 'tc-accounts';
import { getApi } from '../services/api';

/**
 * @static
 * @desc Creates an action that loads Topcoder user profile from v3 API.
 * @param {String} userTokenV3 v3 authentication token.
 * @return {Action}
 */
function loadProfileDone(userTokenV3) {
  if (!userTokenV3) return Promise.resolve(null);
  const user = decodeToken(userTokenV3);
  const api = getApi('V3', userTokenV3);
  const apiV5 = getApi('V5', userTokenV3);
  return Promise.all([
    api.get(`/members/${user.handle}`)
      .then(res => res.json()).then(res => (res.result.status === 200 ? res.result.content : {})),
    apiV5.get(`/groups?memberId=${user.userId}&membershipType=user`)
      .then(res => (res.ok ? res.json() : new Error(res.statusText))),
  ]).then(([profile, groups]) => ({ ...profile, groups }));
}

/**
 * @static
 * @desc Creates an action that sets Topcoder v2 authentication token.
 * @param {String} tokenV2 Topcoder v2 authentication token.
 * @return {Action}
 */
function setTcTokenV2(tokenV2) {
  return tokenV2;
}

/**
 * @static
 * @desc Creates an action that decodes Topcoder v3 authentication token,
 * to get user object, and then writes both the token and the user object into
 * Redux store.
 * @param {String} tokenV3 Topcoder v3 authentication token.
 * @return {Action}
 */
function setTcTokenV3(tokenV3) {
  return tokenV3;
}

export default createActions({
  AUTH: {
    LOAD_PROFILE: loadProfileDone,
    SET_TC_TOKEN_V2: setTcTokenV2,
    SET_TC_TOKEN_V3: setTcTokenV3,
  },
});
