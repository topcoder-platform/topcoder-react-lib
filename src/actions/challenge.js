/**
 * @module "actions.challenge"
 * @desc Actions related to Topcoder challenges APIs.
 */

import _ from 'lodash';
import { config } from 'topcoder-react-utils';
import { createActions } from 'redux-actions';
import { getService as getChallengesService } from '../services/challenges';
import { getService as getSubmissionService } from '../services/submissions';
import { getApi } from '../services/api';

/**
 * @static
 * @desc Creates an action that drops from Redux store all checkpoints loaded
 *  before.
 * @return {Action}
 */
function dropCheckpoints() {}

/**
 * @static
 * @desc Creates an action that drops from Redux store all challenge results
 *  loaded before.
 * @return {Action}
 */
function dropResults() {}

/**
 * @static
 * @desc Creates an action that signals beginning of challenge details loading.
 * @param {Number|String} challengeId Challenge ID
 * @return {Action}
 */
function getDetailsInit(challengeId) {
  return _.toString(challengeId);
}

/**
 * @static
 * @desc Creates an action that loads challenge details.
 * @param {Number|String} challengeId Challenge ID.
 * @param {String} tokenV3 Topcoder v3 auth token.
 * @param {String} tokenV2 Topcoder v2 auth token.
 * @return {Action}
 */
function getDetailsDone(challengeId, tokenV3, tokenV2) {
  const service = getChallengesService(tokenV3, tokenV2);
  const v3Promise = service.getChallengeDetails(challengeId);
  return v3Promise;
}

/**
 * @static
 * @desc Creates an action that signals beginning of user submissions loading.
 * @param {String} challengeId Challenge ID.
 * @return {Action}
 */
function getSubmissionsInit(challengeId) {
  /* As a safeguard, we enforce challengeId to be string (in case somebody
   * passes in a number, by mistake). */
  return _.toString(challengeId);
}

/**
 * @static
 * @desc Creates an action that loads user's submissions to the specified
 * challenge.
 * @param {String} challengeId Challenge ID.
 * @param {String} tokenV2  Topcoder auth token v2.
 * @return {Action}
 */
function getSubmissionsDone(challengeId, tokenV2) {
  return getApi('V2', tokenV2)
    .fetch(`/challenges/submissions/${challengeId}/mySubmissions`)
    .then(response => response.json())
    .then(response => ({
      challengeId: _.toString(challengeId),
      submissions: response.submissions,
    }))
    .catch((error) => {
      const err = { challengeId: _.toString(challengeId), error };
      throw err;
    });
}

/**
 * @static
 * @desc Creates an action that signals beginning of Marathon Match submissions loading.
 * @param {String} challengeId Challenge ID.
 * @return {Action}
 */
function getMMSubmissionsInit(challengeId) {
  /* As a safeguard, we enforce challengeId to be string (in case somebody
   * passes in a number, by mistake). */
  return _.toString(challengeId);
}


/**
 * @static
 * @desc Creates an action that loads Marathon Match submissions to the specified
 * challenge.
 * @param {String} challengeId Challenge ID.
 * @param {Array} registrants The array of register.
 * @param {String} tokenV3  Topcoder auth token v3.
 * @return {Action}
 */
async function getMMSubmissionsDone(challengeId, registrants, tokenV3) {
  const submissionsService = getSubmissionService(tokenV3);
  const submissions = await submissionsService.getSubmissions(challengeId);
  return { challengeId, submissions, tokenV3 };
}

/**
 * @static
 * @desc Creates an action that signals beginning of registration for a
 * challenge.
 * @return {Action}
 */
function registerInit() {
}

/**
 * @static
 * @desc Creates an action that registers user for a challenge.
 * @param {Object} auth An object that holds auth tokens. You can directly pass
 *  here the `auth` segment of Redux store.
 * @param [auth.tokenV2]{String} Topcoder auth token v2.
 * @param [auth.tokenV3]{String} Topcoder auth token v3.
 * @param {String} challengeId Challenge ID.
 * @return {Action}
 */
function registerDone(auth, challengeId) {
  return getChallengesService(auth.tokenV3)
    .register(challengeId)
    /* As a part of registration flow we silently update challenge details,
     * reusing for this purpose the corresponding action handler. */
    // Uses a delay to allow API time to update
    .then(() => new Promise(
      resolve => setTimeout(
        () => resolve(getDetailsDone(challengeId, auth.tokenV3, auth.tokenV2)),
        config.CHALLENGE_DETAILS_REFRESH_DELAY,
      ),
    ));
}

/**
 * @static
 * @desc Creates an action that signals beginning of user unregistration from a
 *  challenge.
 * @return {Action}
 */
function unregisterInit() {}

/**
 * @static
 * @desc Creates an action that unregisters user from a challenge.
 * @param {Object} auth Object that holds Topcoder auth tokens.
 * @param {String} [auth.tokenV2] v2 token.
 * @param {String} [auth.tokenV3] v3 token.
 * @param {String} challengeId Challenge ID.
 * @return {Action}
 */
function unregisterDone(auth, challengeId) {
  return getChallengesService(auth.tokenV3)
    .unregister(challengeId)
    /* As a part of unregistration flow we silently update challenge details,
     * reusing for this purpose the corresponding action handler. */
    // Uses a delay to allow API time to update
    .then(() => new Promise(
      resolve => setTimeout(
        () => resolve(getDetailsDone(challengeId, auth.tokenV3, auth.tokenV2)),
        config.CHALLENGE_DETAILS_REFRESH_DELAY,
      ),
    ));
}

/**
 * @static
 * @desc Creates an action that signals beginning of challenge results loading.
 * @param {Number|String} challengeId Challenge ID
 * @return {Action}
 */
function loadResultsInit(challengeId) {
  return _.toString(challengeId);
}

/**
 * @static
 * @desc Creates an action that loads challenge results.
 * @param {Object} auth Object that holds Topcoder auth tokens.
 * @param {String} [auth.tokenV2] v2 token.
 * @param {String} [auth.tokenV3] v3 token.
 * @param {Number|String} challengeId Challenge ID. Should match the one passed
 *  in the previous {@link module:actions.challenge.loadResultsInit} call.
 * @param {String} type Challenge type.
 * @return {Action}
 */
function loadResultsDone(auth, challengeId, type) {
  return getApi('V2', auth.tokenV2)
    .fetch(`/${type}/challenges/result/${challengeId}`)
    .then(response => response.json())
    .then(response => ({
      challengeId: _.toString(challengeId),
      results: response.results,
    }));
}

/**
 * @static
 * @desc Creates an action that signals beginning of challenge checkpoints data
 *  loading.
 * @return {Action}
 */
function fetchCheckpointsInit() {}

/**
 * @static
 * @desc Creates an action that loads challenge checkpoints data.
 * @param {String} tokenV2 Topcoder v2 auth token.
 * @param {String} challengeId Challenge ID.
 */
function fetchCheckpointsDone(tokenV2, challengeId) {
  const endpoint = `/design/challenges/checkpoint/${challengeId}`;
  return getApi('V2', tokenV2).fetch(endpoint)
    .then((response) => {
      if (response.status !== 200) {
        throw response.status;
      } else {
        return response.json();
      }
    })
    .then((response) => {
      // Expanded key is used for UI expand/collapse.
      response.checkpointResults.forEach((checkpoint, index) => {
        response.checkpointResults[index].expanded = false;
      });
      return {
        challengeId: Number(challengeId),
        checkpoints: response,
      };
    })
    .catch(error => ({
      error,
      challengeId: Number(challengeId),
    }));
}

/**
 * @static
 * @desc Creates an action that Toggles checkpoint details panel in the Topcoder
 *  Submission Management Page.
 * @todo This is UI action relevant to a specific page in specific app. Must be
 *  moved back to Community App.
 * @param {Number} id Checkpoint ID.
 * @param {Boolean} open Target state: `true` to expand, `false` to collapse the
 *  details.
 * @return {Action}
 */
function toggleCheckpointFeedback(id, open) {
  return { id, open };
}

/**
 * @static
 * @desc Creates an action that signals beginning of challenge details update.
 * @todo No idea, why we have this action. This functionality should be covered
 *  by {@link module:actions.challenge.getDetailsInit} and
 *  {@link module:actions.challenge.getDetailsDone}. We need to refactor this.
 * @param {String} uuid UUID of the operation (the same should be passed into
 *  the corresponding {@link module:actions.challenge.updateChallengeDone}).
 * @return {Action}
 */
function updateChallengeInit(uuid) {
  return uuid;
}

/**
 * @static
 * @desc Creates an action that updates challenge details.
 * @todo No idea, why we have this action. This functionality should be covered
 *  by {@link module:actions.challenge.getDetailsInit} and
 *  {@link module:actions.challenge.getDetailsDone}. We need to refactor this.
 * @param {String} uuid Operation UUID. Should match the one passed into the
 *  previous {@link module:actions.challenge.updateChallengeInit} call.
 * @param {Object} challenge Challenge data.
 * @param {String} tokenV3 Topcoder v3 auth token.
 * @return {Action}
 */
function updateChallengeDone(uuid, challenge, tokenV3) {
  return getChallengesService(tokenV3).updateChallenge(challenge)
    .then(res => ({ uuid, res }));
}

/**
 * @static
 * @desc Creates an action that signals beginning of getting count of user's active challenges.
 * @return {Action}
 */
function getActiveChallengesCountInit() {}

/**
 * @static
 * @desc Creates an action that gets count of user's active challenges from the backend.
 * @param {String} handle Topcoder user handle.
 * @param {String} tokenV3 Optional. Topcoder auth token v3. Without token only
 *  public challenges will be counted. With the token provided, the action will
 *  also count private challenges related to this user.
 * @return {Action}
 */
function getActiveChallengesCountDone(handle, tokenV3) {
  return getChallengesService(tokenV3).getActiveChallengesCount(handle);
}

/**
 * @static
 * @desc Creates an action that gets submission information by submission id
 * @param {String} submissionId The submission id
 * @return {Action}
 */
function getSubmissionInformationInit(submissionId) {
  return _.toString(submissionId);
}

/**
 * @static
 * @desc Creates an action that gets submission information from the backend.
 * @param {String} submissionId The submission id
 * @param {String} tokenV3 Topcoder auth token v3.
 * @return {Action}
 */
function getSubmissionInformationDone(submissionId, tokenV3) {
  return getSubmissionService(tokenV3)
    .getSubmissionInformation(submissionId)
    .then(response => ({
      submissionId, submission: response,
    }));
}

export default createActions({
  CHALLENGE: {
    DROP_CHECKPOINTS: dropCheckpoints,
    DROP_RESULTS: dropResults,
    FETCH_CHECKPOINTS_INIT: fetchCheckpointsInit,
    FETCH_CHECKPOINTS_DONE: fetchCheckpointsDone,
    GET_DETAILS_INIT: getDetailsInit,
    GET_DETAILS_DONE: getDetailsDone,
    GET_SUBMISSIONS_INIT: getSubmissionsInit,
    GET_SUBMISSIONS_DONE: getSubmissionsDone,
    LOAD_RESULTS_INIT: loadResultsInit,
    LOAD_RESULTS_DONE: loadResultsDone,
    REGISTER_INIT: registerInit,
    REGISTER_DONE: registerDone,
    TOGGLE_CHECKPOINT_FEEDBACK: toggleCheckpointFeedback,
    UNREGISTER_INIT: unregisterInit,
    UNREGISTER_DONE: unregisterDone,
    UPDATE_CHALLENGE_INIT: updateChallengeInit,
    UPDATE_CHALLENGE_DONE: updateChallengeDone,
    GET_ACTIVE_CHALLENGES_COUNT_INIT: getActiveChallengesCountInit,
    GET_ACTIVE_CHALLENGES_COUNT_DONE: getActiveChallengesCountDone,
    GET_MM_SUBMISSIONS_INIT: getMMSubmissionsInit,
    GET_MM_SUBMISSIONS_DONE: getMMSubmissionsDone,
    GET_SUBMISSION_INFORMATION_INIT: getSubmissionInformationInit,
    GET_SUBMISSION_INFORMATION_DONE: getSubmissionInformationDone,
  },
});
