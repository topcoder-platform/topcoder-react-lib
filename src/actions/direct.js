/**
 * @module "actions.direct"
 * @desc Actions related to Direct API: access to projects, billing accounts,
 * copilot operations, and other similar stuff is handled by them.
 */

import _ from 'lodash';
import { createActions } from 'redux-actions';
import { getService } from '../services/direct';

/**
 * @static
 * @desc Creates an action that drops out of Redux store all Direct-related
 *  data, loaded by other actions from this module, and also cancels any pending
 *  loading operations.
 * @return {Action}
 */
function dropAll() {
  return null;
}

/**
 * @static
 * @desc Creates an action that signals beginning of project details loading.
 * @param {Number} projectId Project ID.
 * @return {Action}
 */
function getProjectDetailsInit(projectId) {
  return projectId;
}

/**
 * @static
 * @desc Creates an action that loads project details.
 * @param {Number} projectId Project ID.
 * @param {String} tokenV3 Topcoder auth token v3.
 * @return {Action}
 */
function getProjectDetailsDone(projectId, tokenV3) {
  return getService(tokenV3).getProjectDetails(projectId);
}

/**
 * @static
 * @desc Creates an action that signals beginning of project permissions
 *  loading.
 * @param {Number|String} projectId Project ID.
 * @return {Action}
 */
function getProjectPermissionsInit(projectId) {
  return _.toNumber(projectId);
}

/**
 * @static
 * @desc Creates an action that loads project permissions.
 * @param {Number|String} projectId Project ID.
 * @param {String} tokenV3 Topcoder v3 auth token.
 * @return {Action}
 */
function getProjectPermissionsDone(projectId, tokenV3) {
  return getService(tokenV3).getProjectPermissions(projectId)
    .then(permissions => ({ permissions, projectId }));
}

/**
 * @static
 * @desc Creates an action that signals beginning of loading the projects
 *  related with a user.
 * @param {String} tokenV3 Topcoder v3 auth token of the user for who we load
 *  the projects.
 * @return {Action}
 */
function getUserProjectsInit(tokenV3) {
  return tokenV3;
}

/**
 * @static
 * @desc Creates an action that loads projects related to a user.
 * @param {String} tokenV3 Topcoder auth token v3.
 * @param {Boolean} hasActiveBillingAccount Optional. Defaults to false.
 *  Whether only projects with active billing accounts should be included
 *  into the results.
 * @return {Action}
 */
async function getUserProjectsDone(tokenV3, hasActiveBillingAccount) {
  const projects = await getService(tokenV3).getUserProjects();
  if (hasActiveBillingAccount) {
    _.remove(projects, proj => !proj.billingAccountId || proj.billingAccountId === 0);
  }
  return { tokenV3, projects };
}

export default createActions({
  DIRECT: {
    DROP_ALL: dropAll,
    GET_PROJECT_DETAILS_INIT: getProjectDetailsInit,
    GET_PROJECT_DETAILS_DONE: getProjectDetailsDone,
    GET_PROJECT_PERMISSIONS_INIT: getProjectPermissionsInit,
    GET_PROJECT_PERMISSIONS_DONE: getProjectPermissionsDone,
    GET_USER_PROJECTS_INIT: getUserProjectsInit,
    GET_USER_PROJECTS_DONE: getUserProjectsDone,
  },
});
