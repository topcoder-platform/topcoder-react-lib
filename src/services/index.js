/**
 * Export services.
 */
import * as api from './api';
import * as billing from './billing';
import * as challenge from './challenges';
import * as direct from './direct';
import * as groups from './groups';
import * as members from './members';
import * as terms from './terms';
import * as communities from './communities';
import * as reviewOpportunities from './reviewOpportunities';
import * as userSetting from './user-settings';
import * as user from './user';
import * as lookup from './lookup';

export const services = {
  api,
  billing,
  terms,
  communities,
  challenge,
  direct,
  groups,
  members,
  user,
  userSetting,
  reviewOpportunities,
  lookup,
};

export default undefined;
