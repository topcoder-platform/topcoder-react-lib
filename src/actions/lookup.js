/**
 * @module "actions.lookup"
 * @desc Actions related to lookup data.
 */

import { createActions } from 'redux-actions';
import { getService } from '../services/lookup';

/**
 * @static
 * @desc Creates an action that signals beginning of getting all skill tags.
 * @return {Action}
 */
function getSkillTagsInit() {}

/**
 * @static
 * @desc Creates an action that gets all skill tags.
 * @return {Action}
 */
function getSkillTagsDone() {
  const params = {
    domain: 'SKILLS',
    status: 'APPROVED',
  };
  return getService().getTags(params);
}

export default createActions({
  LOOKUP: {
    GET_SKILL_TAGS_INIT: getSkillTagsInit,
    GET_SKILL_TAGS_DONE: getSkillTagsDone,
  },
});
