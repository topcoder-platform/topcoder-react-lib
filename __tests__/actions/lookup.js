import { redux } from 'topcoder-react-utils';

import * as LookupService from 'services/lookup';
import actions from 'actions/lookup';

const tag = {
  domain: 'SKILLS',
  id: 251,
  name: 'Jekyll',
  status: 'APPROVED',
};

// Mock services
const mockLookupService = {
  getTags: jest.fn().mockReturnValue(Promise.resolve([tag])),
};
LookupService.getService = jest.fn().mockReturnValue(mockLookupService);

test('Module exports', () => expect(actions).toMatchSnapshot());

test('lookup.getSkillTagsInit', async () => {
  const actionResult = actions.lookup.getSkillTagsInit();
  expect(actionResult).toMatchSnapshot();
});

test('lookup.getSkillTagsDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.lookup.getSkillTagsDone());
  expect(actionResult).toMatchSnapshot();
  expect(mockLookupService.getTags).toBeCalled();
});
