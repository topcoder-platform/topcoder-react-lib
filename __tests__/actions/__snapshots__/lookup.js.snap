// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module exports 1`] = `
Object {
  "lookup": Object {
    "getSkillTagsDone": [Function],
    "getSkillTagsInit": [Function],
  },
}
`;

exports[`lookup.getSkillTagsDone 1`] = `
Object {
  "payload": Array [
    Object {
      "domain": "SKILLS",
      "id": 251,
      "name": "Jekyll",
      "status": "APPROVED",
    },
  ],
  "type": "LOOKUP/GET_SKILL_TAGS_DONE",
}
`;

exports[`lookup.getSkillTagsInit 1`] = `
Object {
  "type": "LOOKUP/GET_SKILL_TAGS_INIT",
}
`;
